tests/expressions_window/tests.py

```diff
@@ -0,0 +1,658 @@
+        all of the different types of window functions as well as partitioning
+        and frames.
+
+        The test data was partially inspired by the documentation in
+        PostgreSQL, and from some third-party material that explained
+        window functions.
+        """
+        WindowTestModel.objects.create(name='Jones', salary=45000, department='Accounting',
+                                       hiredate=datetime.datetime(2005, 11, 1))
+        WindowTestModel.objects.create(name='Williams', salary=37000, department='Accounting',
+                                       hiredate=datetime.datetime(2009, 6, 1))
+        WindowTestModel.objects.create(name='Jenson', salary=45000, department='Accounting',
+                                       hiredate=datetime.datetime(2008, 4, 1))
+        WindowTestModel.objects.create(name='Adams', salary=50000, department='Accounting',
+                                       hiredate=datetime.datetime(2013, 7, 1))
+
+        WindowTestModel.objects.create(name='Smith', salary=55000, department='Sales',
+                                       hiredate=datetime.datetime(2007, 6, 1))
+        WindowTestModel.objects.create(name='Brown', salary=53000, department='Sales',
+                                       hiredate=datetime.datetime(2009, 9, 1))
+
+        WindowTestModel.objects.create(name='Johnson', salary=40000, department='Marketing',
+                                       hiredate=datetime.datetime(2012, 3, 1))
+        WindowTestModel.objects.create(name='Smith', salary=38000, department='Marketing',
+                                       hiredate=datetime.datetime(2009, 10, 1))
+
+        WindowTestModel.objects.create(name='Wilkinson', salary=60000, department='IT',
+                                       hiredate=datetime.datetime(2011, 3, 1))
+        WindowTestModel.objects.create(name='Moore', salary=34000, department='IT',
+                                       hiredate=datetime.datetime(2013, 8, 1))
+
+        WindowTestModel.objects.create(name='Miller', department='Management', salary=100000,
+                                       hiredate=datetime.datetime(2005, 6, 1))
+        WindowTestModel.objects.create(name='Johnson', department='Management', salary=80000,
+                                       hiredate=datetime.datetime(2005, 6, 1))
+
+    @skipUnlessSupportsFunction('DenseRank')
+    def test_invalid_filter(self):
+        with self.assertRaisesMessage(NotSupportedError, 'Window is disallowed in the filter-clause'):
+            WindowTestModel.objects.annotate(dense_rank=Window(
+                expression=DenseRank())
+            ).filter(dense_rank__gte=10)
+
+    @skipUnlessSupportsFunction('DenseRank')
+    def test_dense_rank(self):
+        qs = WindowTestModel.objects.annotate(dense_rank=Window(
+            expression=DenseRank(),
+            partition_by=F('department'),
+            order_by='-salary')
```

Comment:
We have two employees in the same department (_"Accounting"_) and with the same salary (_"45000"_) i.e. _"Jenson_", _"Johnson_". Result is non-deterministic, it depends on the db backend default ordering, that seems to be different for each database. IMO the best solution is to add option to pass multiple field names (like in [`models/query.py`](https://github.com/django/django/blob/master/django/db/models/query.py#L926)) in `order_by` attribute, e.g.
```python
    def test_dense_rank(self):
        qs = WindowTestModel.objects.annotate(dense_rank=Window(
            expression=DenseRank(),
            partition_by=F('department'),
            order_by=['-salary', 'name'])
        ).order_by('department')
```
Ordering is also an issue in others tests.